### PR TITLE
dbr quant: add limited support for `torch.nn.ModuleList`

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#70372 dbr quant: add limited support for `torch.nn.ModuleList`**
* #70349 dbr quant support for custom leaf modules, part 3/x
* #70335 dbr quant: support for custom leaf modules, part 2/x
* #70330 dbr quant: support for custom leaf modules, part 1/x
* #70329 dbr quant: unbreak case when child module not returning any outputs
* #70328 dbr quant: support functional linear without bias
* #70257 dbr quant: support dynamic linear
* #70256 dbr quant: do not attach auto_quant_state to observers

Summary:

Enables basic support for `torch.nn.ModuleList` in DBR quant
by stopping it from being a leaf.  For now, we
require the user to check for `AutoQuantizationState` if they are
looping over the contents without any bounds checking.

In future PRs, we can explore how to solve this without requiring
user code changes.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_module_list
```

Differential Revision: [D33302329](https://our.internmc.facebook.com/intern/diff/D33302329)

### CODE REVIEW

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33302329).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33302329).

File: test/quantization/dbr/test_quantize_dbr.py [hunk 1]
```diff
@@ -704,6 +705,33 @@ def forward(self, x):
         FileCheck().check_count("aten::add", 1, exactly=True).run(mqt.m2.graph)
         FileCheck().check_count("quantized::add", 0, exactly=True).run(mqt.m2.graph)
 
+    def test_module_list(self):
```
Comment (jerryzh168) [inline test/quantization/dbr/test_quantize_dbr.py:708]: nit: just to confirm, do we have a test for Sequential as well? also for completeness do we want to add support for ModuleDict?

**Review by jerryzh168 — APPROVED** [review]

File: test/quantization/dbr/test_quantize_dbr.py (hunk 1 above)
Comment (vkuzo) [reply test/quantization/dbr/test_quantize_dbr.py:708]: yes, sequential is already tested.  ModuleDict could be added in a future PR, it's just not needed for the current model we are working on.

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33302329).

**Pull request closed without merging**